lib/on-contract-event.js

```diff
@@ -0,0 +1,100 @@
+// eth_getFilterChanges isn't fully supported by lotus, see
+// https://filecoinproject.slack.com/archives/CRK2LKYHW/p1705404161018769
+// Therefore we need to poll for events
+
+import { ethers } from 'ethers'
+import timers from 'node:timers/promises'
+
+const MIN_POLL_INTERVAL = 10_000
+
+const makeRequest = async ({
+  contract,
+  address,
+  rpcUrl,
+  rpcHeaders,
+  lastBlockNumber
+}) => {
+  const req = {
+    jsonrpc: '2.0',
+    id: 1,
+    method: 'eth_getLogs',
+    params: [
+      {
+        address,
+        fromBlock: ethers.toBeHex(lastBlockNumber)
+      }
+    ]
+  }
+  const res = await fetch(rpcUrl, {
+    method: 'POST',
+    headers: {
+      ...rpcHeaders,
+      'content-type': 'application/json'
+    },
+    body: JSON.stringify(req)
+  })
+  if (!res.ok) {
+    throw new Error(
+      `Cannot fetch event log. JSON RPC error ${res.status}\n${await res.text()}`
+    )
+  }
+  const body = await res.json()
+  if (body.error) {
+    throw new Error(`Cannot fetch event log. JSON RPC error: ${body.error}`)
+  }
+
+  const events = body.result.map(log => contract.interface.parseLog(log))
+  if (body.result.length > 0) {
+    lastBlockNumber = body.result
+      .map(r => ethers.getBigInt(r.blockNumber))
+      .sort()
+      .pop()
```

Comment:
Sorting the entire array is `O(N.logN)`, finding the maximum item can be implemented in `O(N)` where `N=body.result.length)`.

Not a big deal if we are confident that the number of events will stay low.

Reply:
I don't think this matters in this context, and is therefore a premature optimization. In a polling interval we will at most see two new events